mlir/unittests/Analysis/Presburger/SimplexTest.cpp

```diff
@@ -0,0 +1,219 @@
+//===- SimplexTest.cpp - Tests for Simplex ----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Analysis/Presburger/Simplex.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+namespace mlir {
+
+TEST(SimplexTest, Empty) {
+  Simplex tab(2);
+  // (u - v) >= 0
+  tab.addInequality({1, -1, 0});
+  // (u - v) <= -1
+  EXPECT_FALSE(tab.isEmpty());
+  tab.addInequality({-1, 1, -1});
+  EXPECT_TRUE(tab.isEmpty());
+}
+
+TEST(SimplexTest, isMarkedRedundant_no_var_ge_zero) {
+  Simplex tab(0);
+  tab.addInequality({0}); // 0 >= 0
+
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+  EXPECT_TRUE(tab.isMarkedRedundant(0));
+}
+
+TEST(SimplexTest, isMarkedRedundant_no_var_eq) {
+  Simplex tab(0);
+  tab.addEquality({0}); // 0 == 0
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+  EXPECT_TRUE(tab.isMarkedRedundant(0));
+}
+
+TEST(SimplexTest, isMarkedRedundant_pos_var_eq) {
+  Simplex tab(1);
+  tab.addEquality({1, 0}); // x == 0
+
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+  EXPECT_FALSE(tab.isMarkedRedundant(0));
+}
+
+TEST(SimplexTest, isMarkedRedundant_zero_var_eq) {
+  Simplex tab(1);
+  tab.addEquality({0, 0}); // 0x == 0
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+  EXPECT_TRUE(tab.isMarkedRedundant(0));
+}
+
+TEST(SimplexTest, isMarkedRedundant_neg_var_eq) {
+  Simplex tab(1);
+  tab.addEquality({-1, 0}); // -x == 0
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+  EXPECT_FALSE(tab.isMarkedRedundant(0));
+}
+
+TEST(SimplexTest, isMarkedRedundant_pos_var_ge) {
+  Simplex tab(1);
+  tab.addInequality({1, 0}); // x >= 0
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+  EXPECT_FALSE(tab.isMarkedRedundant(0));
+}
+
+TEST(SimplexTest, isMarkedRedundant_zero_var_ge) {
+  Simplex tab(1);
+  tab.addInequality({0, 0}); // 0x >= 0
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+  EXPECT_TRUE(tab.isMarkedRedundant(0));
+}
+
+TEST(SimplexTest, isMarkedRedundant_neg_var_ge) {
+  Simplex tab(1);
+  tab.addInequality({-1, 0}); // x <= 0
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+  EXPECT_FALSE(tab.isMarkedRedundant(0));
+}
+
+TEST(SimplexTest, isMarkedRedundant) {
+  Simplex tab(3);
+
+  tab.addEquality({-1, 0, 1, 0});     // u = w
+  tab.addInequality({-1, 16, 0, 15}); // 15 - (u - 16v) >= 0
+  tab.addInequality({1, -16, 0, 0});  //      (u - 16v) >= 0
+
+  tab.detectRedundant();
+  ASSERT_FALSE(tab.isEmpty());
+
+  for (unsigned i = 0; i < tab.numberConstraints(); ++i)
+    EXPECT_FALSE(tab.isMarkedRedundant(i)) << "i = " << i << "\n";
+}
+
+TEST(SimplexTest, isMarkedRedundant_a) {
```

Comment:
What the suffix "_a" stands for?

Reply:
Not sure. I gave all the tests better names.